### PR TITLE
[FIX] clipboard: fix behaviour when selection is updated before the paste

### PR DESCRIPTION
**Pull request by khpa-odoo**
## Description:

Cut/Copy:
After a cut/copy operation, if the selection is updated before paste,  the content was pasted w/o any style. Also, the cut zone was not destroyed in case of cut operation.

This fix solves the problem by using the cells stored inside the Clipboard Plugin instead of the current state of the cells in the grid.

Also, this fix restricts pasting the content multiple times after a cut operation and clears the clipboard after a cut and single paste as well.

Only Format:
Copy external content and do 'Paste' or 'Paste-Only Value' on a cell. Now,  performing a 'Paste-Only Format' on a different cell would paste the content as well.

This fix solves the problem by handling content in case the pasteOption is 'onlyFormat'.

Odoo task ID : [3135623](https://www.odoo.com/web#id=3135623&cids=2&menu_id=4720&action=333&active_id=2328&model=project.task&view_type=form)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: tests/menu_items_registry.test.ts [hunk 1]
```diff
@@ -173,13 +173,18 @@ describe("Menu Item actions", () => {
     expect(dispatch).toHaveBeenCalledWith("PASTE_FROM_OS_CLIPBOARD", {
       target: target("A1"),
       text,
+      pasteOption: "onlyValue",
     });
   });
 
-  test("Edit -> paste_special -> paste_special_format", () => {
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:179]: Why change this test ? Only "paste_special_value from OS clipboard" should have been changed right ?

File: src/plugins/ui_stateful/clipboard.ts [hunk 2]
```diff
@@ -157,9 +156,10 @@ export class ClipboardPlugin extends UIPlugin {
       }
       case "PASTE_FROM_OS_CLIPBOARD":
         this.state = new ClipboardOsState(cmd.text, this.getters, this.dispatch, this.selection);
-        this.state.paste(cmd.target);
+        this.state.paste(cmd.target, { pasteOption: cmd.pasteOption });
         this.lastPasteState = this.state;
         this.status = "invisible";
+        this.allowPaste = true;
```
Comment (hokolomopo) [inline src/plugins/ui_stateful/clipboard.ts:162]: Is that useful ?

File: src/plugins/ui_stateful/clipboard.ts [hunk 3]
```diff
@@ -87,17 +87,16 @@ export class ClipboardPlugin extends UIPlugin {
         const zones = ("target" in cmd && cmd.target) || this.getters.getSelectedZones();
         this.state = this.getClipboardState(zones, cmd.type);
         this.status = "visible";
+        this.allowPaste = true;
         break;
       case "PASTE":
-        if (!this.state) {
+        if (!this.state || !this.allowPaste) {
           break;
         }
```
Comment (hokolomopo) [inline src/plugins/ui_stateful/clipboard.ts:95]: this should be in the allowDispatch

File: src/plugins/ui_stateful/clipboard.ts [hunk 4]
```diff
@@ -87,17 +87,16 @@ export class ClipboardPlugin extends UIPlugin {
...
+        if (this.state?.operation === "CUT") this.allowPaste = false;
```
Comment (hokolomopo) [inline src/plugins/ui_stateful/clipboard.ts:96]:
```suggestion
        if (this.state?.operation === "CUT") {
            this.allowPaste = false;
        }
```
some peoples in the teeam don't like one-line condition, so try to avoid them.

File: tests/plugins/clipboard.test.ts [hunk 5]
```diff
@@ -190,16 +192,140 @@ describe("clipboard", () => {
...
+    pasteFromOSClipboard(model, "C3", content, "onlyFormat");
+    expect(getCellContent(model, "C3")).toBe("");
+  });
+
+  test("Internal copy followed by OS copy should not bring paste format from internal copy", async () => {
+    const model = new Model();
+    setCellContent(model, "A1", "a1");
+    setStyle(model, "A1", { fillColor: "#FA0000" });
+    copy(model, "A1");
+    expect(getStyle(model, "A1")).toEqual({ fillColor: "#FA0000" });
```
Comment (hokolomopo) [inline tests/plugins/clipboard.test.ts:273]: This is useless no?

File: src/helpers/clipboard/clipboard_os_state.ts [hunk 6]
```diff
@@ -45,10 +45,16 @@ export class ClipboardOsState extends ClipboardCellsAbstractState {
     this.addMissingDimensions(numberOfCols, numberOfRows, activeCol, activeRow);
     for (let i = 0; i < values.length; i++) {
       for (let j = 0; j < values[i].length; j++) {
+        let content: string;
+        if (clipboardOption?.pasteOption === "onlyFormat") {
+          content = "";
```
Comment (hokolomopo) [inline src/helpers/clipboard/clipboard_os_state.ts:50]: This should be undefined not an empty string, we don't want to clear the cell when pasting only the format. 
Or better we can just do an early return for `onlyFormat`, we don't need to loop on the cells and dispatch UPDATE_CELL  to do nothing on them.

**Review by hokolomopo — COMMENTED** [review]
You should make sure taht the tests actually test what you want, here some of the tests you created don't really test the issue. 

A good way to do that is to launch the test without your fix; they should fail here, and only work after your fix.

File: tests/menu_items_registry.test.ts (hunk 1 above)
Comment (khpa-odoo) [reply tests/menu_items_registry.test.ts:179]: Since we've changed the case to be called for `onlyFormat`, this test case would expect paste_special_format action to dispatch "PASTE_FROM_OS_CLIPBOARD" instead of "PASTE". Hence the change :)

File: src/plugins/ui_stateful/clipboard.ts (hunk 2 above)
Comment (khpa-odoo) [reply src/plugins/ui_stateful/clipboard.ts:164]: If we cut something, 'this.allowPaste' will be set to false. When we paste external text now, the first paste works perfectly. When you try to paste that text again, "PASTE" is called, and 'this.allowPaste = false' prevents the content from being pasted. As a result, we set 'this.allowPaste = true' to allow repeated pastes when using external content.

File: tests/plugins/clipboard.test.ts (hunk 5 above)
Comment (khpa-odoo) [reply tests/plugins/clipboard.test.ts:273]: Main objective behind this test case was to check an edge case with "onlyFormat". Once we copy something internally, & followed we make an external copy, it was observed that upon special pasting format, it was bringing the format of first internal copy, which is obviously wrong. Hence style and content is set beforehand. Although your point does stay that pasteFromOSClipboard method will disregard the above anyway, But then I don't really can figure out how to recreate this issue in test case.

File: src/helpers/clipboard/clipboard_os_state.ts [hunk 7]
```diff
@@ -45,12 +45,16 @@ export class ClipboardOsState extends ClipboardCellsAbstractState {
...
       for (let j = 0; j < values[i].length; j++) {
-        this.dispatch("UPDATE_CELL", {
-          row: activeRow + i,
-          col: activeCol + j,
-          content: values[i][j],
-          sheetId,
-        });
+        if (clipboardOption?.pasteOption === "onlyFormat") {
+          return;
+        } else {
```
Comment (hokolomopo) [inline src/helpers/clipboard/clipboard_os_state.ts:50]: place the `if() return` before the loop.

File: src/plugins/ui_stateful/clipboard.ts [hunk 8]
```diff
@@ -52,19 +52,31 @@ export class ClipboardPlugin extends UIPlugin {
 
   allowDispatch(cmd: LocalCommand): CommandResult {
     switch (cmd.type) {
+      case "COPY":
+        this.allowPaste = true;
+        break;
```
Comment (hokolomopo) [inline src/plugins/ui_stateful/clipboard.ts:57]: We should't change the state of the plugin in the `allowDispatch`. It doesn't change much ATM, but we could imagine another plugin that have an allowDispatch on `COPY`. We could then have the situation:

- ClipboardPlugin allowDispatch COPY, set `allowPaste` to true
- Random plugin block the COPY in its allowDispatch.
- In this case the state of the ClipboardPlugin is wrong, because allowPatch = true, but we didn't actually copy anything.

You did well adding the `CommandResult.RepeatitivePasteAfterCut` in the allowDispatch. But the changes on the allowPaste should stay in the `handle`.

File: src/types/commands.ts [hunk 9]
```diff
@@ -1169,6 +1170,7 @@ export const enum CommandResult {
   SplitWillOverwriteContent,
   NoSplitSeparatorInSelection,
   NoActiveSheet,
+  RepeatitivePasteAfterCut,
```
Comment (hokolomopo) [inline src/types/commands.ts:1173]:
```suggestion
  RepeatitedPasteAfterCut,
```

File: tests/menu_items_registry.test.ts (hunk 1 above)
Comment (hokolomopo) [reply tests/menu_items_registry.test.ts:180]: We should have a test testing `paste_special_format` calling `PASTE_FROM_OS_CLIPBOARD` and another one calling `PASTE` then :slightly_smiling_face:

File: tests/menu_items_registry.test.ts [hunk 10]
```diff
@@ -174,17 +174,42 @@ describe("Menu Item actions", () => {
...
+        target: target("A1"),
+        text,
+        pasteOption: undefined,
+      });
+    } else {
+      expect(dispatch).toHaveBeenCalledWith("PASTE", {
+        target: target("A1"),
+        pasteOption: undefined,
+      });
+    }
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:210]: The `if/else` have nothing to do here, you are writing a test so you are controlling whether test === content or not. As stated in my previous comment, you should have 2 separate tests : one for PASTE_FROm_OS and one for PASTE.

File: tests/plugins/clipboard.test.ts (hunk 5 above)
Comment (hokolomopo) [reply tests/plugins/clipboard.test.ts:212]: I get your point. The test should be in grid.ts and use the copy/paste function of the grid then. That way you mimick real life behaviour. 

Doing it here is useless, as using `pasteFromOSClipboard` will totally ignore the clipboard state. Doing the test in the grid actually test wheter you call `paste` or `pasteFromOS`, and thus whetehr you paste the internal style or not.

File: tests/plugins/clipboard.test.ts [hunk 11]
```diff
@@ -190,16 +192,77 @@ describe("clipboard", () => {
...
+    expect(getCellContent(model, "C5")).toBe("");
+    expect(getStyle(model, "C5")).toEqual({});
+  });
+
+  test("cannot paste multiple times after cut", () => {
+    const model = new Model();
+    setCellContent(model, "B2", "b2");
+    expect(getCellContent(model, "B2")).toBe("b2");
+    setStyle(model, "B2", { bold: true });
+    expect(getCell(model, "B2")!.style).toEqual({ bold: true });
```
Comment (hokolomopo) [inline tests/plugins/clipboard.test.ts:228]:
```suggestion
    setCellContent(model, "B2", "b2");
    setStyle(model, "B2", { bold: true });
```
we can trust `setCellContent/Style` to do their jobs; we don't need to test every single thing. Same in the other tests.

File: tests/plugins/clipboard.test.ts [hunk 12]
```diff
@@ -190,16 +192,77 @@ describe("clipboard", () => {
...
+    cut(model, "A1");
+    setCellContent(model, "A1", "Orange", sheet1);
+
+    createSheet(model, { sheetId: sheet2 });
+    activateSheet(model, sheet2);
+    deleteSheet(model, sheet1);
+
+    paste(model, "sheet2!A2");
+    expect(getCellContent(model, "A2", sheet2)).toBe("Apple");
+    expect(getCell(model, "A2")!.style).toEqual({ bold: true });
```
Comment (hokolomopo) [inline tests/plugins/clipboard.test.ts:257]: The clipboard still wasn't invalidated  here since the `paste` worked...

**Review by hokolomopo — COMMENTED** [review]
Almost there ! :blush: 

FYI I'll be in vacation until the 30/05.

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 13]
```diff
@@ -481,7 +481,9 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
         .map((cells) => {
           return cells
             .map((c) =>
-              c.cell ? this.getters.getCellText(c.position, this.getters.shouldShowFormulas()) : ""
+              this.getters.shouldShowFormulas()
+                ? c.cell?.content
+                : c.evaluatedCell?.formattedValue || ""
```
Comment (hokolomopo) [inline src/helpers/clipboard/clipboard_cells_state.ts:486]:
```suggestion
              this.getters.shouldShowFormulas()
                ? c.cell?.content || ""
                : c.evaluatedCell?.formattedValue || ""
```

File: src/plugins/ui_stateful/clipboard.ts [hunk 14]
```diff
@@ -60,11 +62,16 @@ export class ClipboardPlugin extends UIPlugin {
...
+          this.status = "invisible";
+          return CommandResult.RepeatedPasteAfterCut;
+        }
         const pasteOption = cmd.pasteOption || (this._isPaintingFormat ? "onlyFormat" : undefined);
         return this.state.isPasteAllowed(cmd.target, { pasteOption });
       case "PASTE_FROM_OS_CLIPBOARD": {
         const state = new ClipboardOsState(cmd.text, this.getters, this.dispatch, this.selection);
-        return state.isPasteAllowed(cmd.target);
+        const result = state.isPasteAllowed(cmd.target, { pasteOption: cmd.pasteOption });
+        return result;
```
Comment (hokolomopo) [inline src/plugins/ui_stateful/clipboard.ts:74]:
```suggestion
        return state.isPasteAllowed(cmd.target, { pasteOption: cmd.pasteOption });
```
nitpick

File: src/plugins/ui_stateful/clipboard.ts [hunk 15]
```diff
@@ -60,11 +62,16 @@ export class ClipboardPlugin extends UIPlugin {
         if (!this.state) {
           return CommandResult.EmptyClipboard;
         }
+        if (!this.allowPaste) {
+          this.status = "invisible";
```
Comment (hokolomopo) [inline src/plugins/ui_stateful/clipboard.ts:66]: Is changing status here useful ?

File: src/plugins/ui_stateful/clipboard.ts [hunk 16]
```diff
@@ -176,6 +188,14 @@ export class ClipboardPlugin extends UIPlugin {
         this.status = "visible";
         break;
       }
+      case "DELETE_SHEET":
+        if (this.state?.operation !== "CUT") {
+          return;
+        }
+        if (this.originSheetId === cmd.sheetId) {
+          this.allowPaste = false;
```
Comment (hokolomopo) [inline src/plugins/ui_stateful/clipboard.ts:196]: should also make the this.status invisible I think

File: tests/components/grid.test.ts [hunk 17]
```diff
@@ -1427,6 +1427,20 @@ describe("Copy paste keyboard shortcut", () => {
...
+  test("cut zone gets cleared on paste if content/style is altered after cut", async () => {
+    setCellContent(model, "A1", "things");
+    setStyle(model, "A1", { bold: true });
+    selectCell(model, "A1");
+    document.body.dispatchEvent(getClipboardEvent("cut", clipboardData));
+    setCellContent(model, "A1", "new content");
+    setStyle(model, "A1", { bold: false });
+    selectCell(model, "A2");
+    document.body.dispatchEvent(getClipboardEvent("paste", clipboardData));
+    expect(getCellContent(model, "A2")).toEqual("things");
```
Comment (hokolomopo) [inline tests/components/grid.test.ts:1439]: A2 should contain "new content" no (for CUT) ? Here the changes we made to A1 are totally lost

File: tests/menu_items_registry.test.ts [hunk 18]
```diff
@@ -146,6 +147,37 @@ describe("Menu Item actions", () => {
     expect(getCellContent(model, "A1")).toEqual("");
   });
 
+  test("'Edit -> paste' if copied from grid and content altered before paste", async () => {
+    setCellContent(model, "A1", "a1");
+    doAction(["edit", "copy"], env); // first copy from grid
+    setCellContent(model, "A1", "os clipboard");
+    selectCell(model, "C3");
+    doAction(["edit", "paste"], env);
+    await nextTick();
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:156]: nextTick() useless I think

File: tests/menu_items_registry.test.ts [hunk 19]
```diff
@@ -146,6 +147,37 @@ describe("Menu Item actions", () => {
...
+
+  test("Internal copy followed by OS copy should not bring paste format from internal copy", async () => {
+    setCellContent(model, "C1", "c1");
+    setStyle(model, "C1", { fillColor: "#FA0000" });
+    selectCell(model, "C1");
+    doAction(["edit", "copy"], env); // first copy from grid
+    await env.clipboard.writeText("Then copy in OS clipboard");
+    selectCell(model, "A1");
+    doAction(["edit", "paste_special", "paste_special_format"], env);
+    await nextTick();
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:172]: nextTick() useless I think

File: tests/menu_items_registry.test.ts [hunk 20]
```diff
@@ -174,17 +206,32 @@ describe("Menu Item actions", () => {
...
       text,
+      pasteOption: "onlyValue",
     });
   });
 
-  test("Edit -> paste_special -> paste_special_format", () => {
+  test("Edit -> paste_special -> paste_special_format", async () => {
+    doAction(["edit", "copy"], env);
     doAction(["edit", "paste_special", "paste_special_format"], env);
+    await nextTick();
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:216]: nextTick() useless I think

File: tests/menu_items_registry.test.ts [hunk 21]
```diff
@@ -174,17 +206,32 @@ describe("Menu Item actions", () => {
...
       target: env.model.getters.getSelectedZones(),
       pasteOption: "onlyFormat",
     });
   });
 
+  test("Edit -> paste_special -> paste_special_format from OS clipboard", async () => {
+    const text = "in OS clipboard";
+    await env.clipboard.writeText(text);
+    doAction(["edit", "paste_special", "paste_special_format"], env);
+    await nextTick();
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:227]: nextTick() useless I think

File: tests/plugins/clipboard.test.ts [hunk 22]
```diff
@@ -190,16 +192,54 @@ describe("clipboard", () => {
...
+    expect(getCell(model, "C2")!.style).toEqual({ bold: true });
+
+    paste(model, "E5");
+    expect(getCell(model, "E5")).toBe(undefined);
+  });
+
+  test("Clipboard should be invalidated when sheet is deleted", async () => {
+    const model = new Model();
+    const sheet1 = model.getters.getActiveSheetId();
+    const sheet2 = "sheet2";
```
Comment (hokolomopo) [inline tests/plugins/clipboard.test.ts:224]:
```suggestion
    const sheet1Id = model.getters.getActiveSheetId();
    const sheet2Id = "sheet2";
```
nitpick

File: tests/plugins/clipboard.test.ts [hunk 23]
```diff
@@ -190,16 +192,54 @@ describe("clipboard", () => {
...
+    createSheet(model, { sheetId: sheet2 });
+
+    setCellContent(model, "A1", "Apple", sheet1);
+    setStyle(model, "A1", { bold: true });
+    cut(model, "A1");
+
+    activateSheet(model, sheet2);
+    deleteSheet(model, sheet1);
+    pasteFromOSClipboard(model, "A2", "Apple");
```
Comment (hokolomopo) [inline tests/plugins/clipboard.test.ts:233]:
```suggestion
    paste(model, "A2");
```
Should be a normal paste.

File: src/plugins/ui_stateful/clipboard.ts [hunk 24]
```diff
@@ -156,10 +165,13 @@ export class ClipboardPlugin extends UIPlugin {
         break;
       }
       case "PASTE_FROM_OS_CLIPBOARD":
-        this.state = new ClipboardOsState(cmd.text, this.getters, this.dispatch, this.selection);
-        this.state.paste(cmd.target);
-        this.lastPasteState = this.state;
-        this.status = "invisible";
+        if (this.allowPaste) {
```
Comment (hokolomopo) [inline src/plugins/ui_stateful/clipboard.ts:168]: `PASTE_FROM_OS_CLIPBOARD` should ignore `this.allowPaste`, since it won't have the correct value. Imagine a situation when we cup/pasted cometing, `this.allowPaste` is false; but then we copy something in another application and try to paste it in the spreadsheet. We don't want to block the paste.

In practice, `PASTE` will be used by grid when the clipboard content was copied from the spreadhseet, `PASTE_FROM_OS_CLIPBOARD` will be used when the clipboard content comes from somewhere else.

**Review by hokolomopo — COMMENTED** [review]
I'm back ! Some comments :)

File: tests/components/grid.test.ts (hunk 17 above)
Comment (khpa-odoo) [reply tests/components/grid.test.ts:1439]: If we go according to the specs of this task, which says PASTE the old values & styles made before CUT/COPY, then I guess this test does solve the purpose right?. Also, this behaviour is same as it was before this fix/task. So do you intend to keep the new values after cut instead? Or keep the old ones as it was?

File: tests/menu_items_registry.test.ts (hunk 18 above)
Comment (khpa-odoo) [reply tests/menu_items_registry.test.ts:156]: It seems (referring to other test cases) after every single doAction performed, we have to add an `await nextTick()` before expecting a command dispatch for that action. I tried removing them as you suggested, but all 4 test cases failed...

Comment (hokolomopo) [reply tests/menu_items_registry.test.ts:156]: Yeah the problem seems that the function `paste` called by the action is async, so we indeed need the `await nextTick();`. Maybe we should change the design and alway sawait the actions, but don't worry about it then.

File: tests/components/grid.test.ts (hunk 17 above)
Comment (hokolomopo) [reply tests/components/grid.test.ts:1439]: When we copy, we should put things in the clipboard, and then paste it as it was copied, without worrying about the changes in the sheet.

Cut/paste is a bit trickier since we delete some sheet content, and we don't want to lose information. Cut can be seen as moving cells from one place to another, so we should use the current value of the sheet.

That's the theory anyway, but looking at the current code we never really did that. I'd say you can leave it like that, and we'll do another task to handle cut correctly.

File: tests/plugins/clipboard.test.ts [hunk 25]
```diff
@@ -190,16 +192,54 @@ describe("clipboard", () => {
...
+    cut(model, "B2");
+    paste(model, "C2");
+    expect(getCellContent(model, "C2")).toBe("b2");
+    expect(getCell(model, "C2")!.style).toEqual({ bold: true });
+
+    paste(model, "E5");
+    expect(getCell(model, "E5")).toBe(undefined);
+  });
+
+  test("Clipboard should be invalidated when sheet is deleted", async () => {
```
Comment (hokolomopo) [inline tests/plugins/clipboard.test.ts:221]:
```suggestion
  test("Cut clipboard should be invalidated when sheet is deleted",  () => {
```

File: tests/plugins/clipboard.test.ts [hunk 26]
```diff
@@ -190,16 +192,54 @@ describe("clipboard", () => {
     expect(getCell(model, "C2")!.style).toEqual({ bold: true });
   });
 
+  test("copying external content & paste-format on a cell will not paste content", async () => {
```
Comment (hokolomopo) [inline tests/plugins/clipboard.test.ts:195]:
```suggestion
  test("copying external content & paste-format on a cell will not paste content",  () => {
```

**Review by hokolomopo — COMMENTED** [review]
Some last nitpick, I'll green the task for the next stage :)

File: tests/menu_items_registry.test.ts (hunk 18 above)
Comment (hokolomopo) [reply tests/menu_items_registry.test.ts:156]: Fixed it in https://github.com/odoo/o-spreadsheet/pull/2522, you should remove the nextTick and await the `doAction` when it's merged and you rebase your branch :)

Comment (khpa-odoo) [reply tests/menu_items_registry.test.ts:156]: Cool :)

File: tests/components/grid.test.ts (hunk 17 above)
Comment (khpa-odoo) [reply tests/components/grid.test.ts:1439]: Okay, I'll leave it as it is then.

Suggestion: We can try to replicate the cut/paste behaviour as in gsheet in the new task.

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 27]
```diff
@@ -481,7 +481,9 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
         .map((cells) => {
           return cells
             .map((c) =>
-              c.cell ? this.getters.getCellText(c.position, this.getters.shouldShowFormulas()) : ""
+              this.getters.shouldShowFormulas()
```
Comment (rrahir) [inline src/helpers/clipboard/clipboard_cells_state.ts:484]:
```suggestion
              this.getters.shouldShowFormulas() && cell?.isFormula
``` 
you should preserve the same logic as `getCellText`. With the current suggestion, if I were to **CUT** a cell with `{content: 1, format: "mm/dd/yyyy"}`  while `this.getters.shouldShowFormulas() === true` I would end up with  `1` in my clipboard, whereas other spreadsheets would yield `12/31/1899`.

I think a test would be welcome here to make sure we don't break tahat behaviour in the future :)

File: src/plugins/ui_stateful/clipboard.ts [hunk 28]
```diff
@@ -43,8 +44,9 @@ export class ClipboardPlugin extends UIPlugin {
   private status: "visible" | "invisible" = "invisible";
   private state?: ClipboardState;
   private lastPasteState?: ClipboardState;
-
+  private allowPaste: boolean = true;
```
Comment (rrahir) [inline src/plugins/ui_stateful/clipboard.ts:47]: After discussing this issue of infinite with the team and PO, we think it would be best (and overall simpler) to empty the OS clipboard as long as the standard clipboard after the **paste** of a **cut** command.  While it is not what google sheet does, i does not make much sense to keep pasting the content of a cut - an action that only makes sense inside the spreadsheet itself.

**Review by rrahir — CHANGES_REQUESTED** [review]
Overall, nice job :+1: 
See my two comments for some modifications I'd like you to add.

On another topic, when you have  several bugs to fix like you do here, please make a dedicated commit for each fix. 2 perks to this:
- it's way easier to review - here I had to guess which change did what in your PR
- there is no loss of information - the commit title here says nothing about the "onlyFormat" fix which has nothing to do with the other fix you are working on

It would be nice if you could split the two changes, (with their respective tests) in separated commits.

File: src/plugins/ui_stateful/clipboard.ts (hunk 28 above)
Comment (khpa-odoo) [reply src/plugins/ui_stateful/clipboard.ts:47]: Hey, I did actually try emptying the clipboard earlier by making its state undefined if operation is 'cut'. But in that case, call would go to paste from os clipboard and there it would be difficult to distinguish whether the content in clipboard is from outside of the sheet or it is from spreadsheet itself.
I've tried multiple approaches to solve this issue but nothing seems to be effective so far. Do you have any suggestions on how this can be achieved?

Comment (hokolomopo) [reply src/plugins/ui_stateful/clipboard.ts:47]: The idea would be to also clear the **OS** Clipboard after a paste from cut.

You cannot do that from the plugin tho, you have to do that from the grid/menu item. Use `env.clipboard.write` to write stuff from the OS clipboard.

**Comment by khpa-odoo** [comment]
Hey @rrahir! I've pushed the changes. Do check out once :)

File: src/plugins/ui_stateful/clipboard.ts (hunk 28 above)
Comment (rrahir) [reply src/plugins/ui_stateful/clipboard.ts:47]: @khpa-odoo I don't think you have addressed this point. the idea would be to completely get rid of the 'allowPaste' boolean that you added and just assume that we empty the clipboard (both the one of the plugin and of the os once we successfully paste after a cut.  In order to do that,you should edit the `src/actions/menu_items_actions.ts  function paste()` to clear the OS clipboard one the paste action was successful in the plugin, but only if we come from a CUT operation, for this you have the getter `isCutOperation`

File: src/helpers/clipboard/clipboard_cells_state.ts (hunk 27 above)
Comment (rrahir) [reply src/helpers/clipboard/clipboard_cells_state.ts]: I don't think I see a test for this

File: tests/components/grid.test.ts [hunk 29]
```diff
@@ -1427,6 +1428,34 @@ describe("Copy paste keyboard shortcut", () => {
...
+    expect(getStyle(model, "A2")).toEqual({ bold: true });
+    expect(getCell(model, "A1")).toBe(undefined);
+  });
+
+  test("Cut and paste of a date format cell, by toggling formula visibility should return date", async () => {
+    model.dispatch("SET_FORMULA_VISIBILITY", { show: true });
+    setCellContent(model, "A1", "1");
+    setCellFormat(model, "A1", "m/d/yyyy");
+    document.body.dispatchEvent(getClipboardEvent("cut", clipboardData));
+    model.dispatch("SET_FORMULA_VISIBILITY", { show: false });
```
Comment (rrahir) [inline tests/components/grid.test.ts:1450]: you can definitely check the content of clipboardDatza, which is what we really are interested in

File: tests/components/grid.test.ts [hunk 30]
```diff
@@ -1427,6 +1428,34 @@ describe("Copy paste keyboard shortcut", () => {
...
+    setCellContent(model, "A1", "new content");
+    setStyle(model, "A1", { bold: false });
+    selectCell(model, "A2");
+    document.body.dispatchEvent(getClipboardEvent("paste", clipboardData));
+    expect(getCellContent(model, "A2")).toEqual("things");
+    expect(getStyle(model, "A2")).toEqual({ bold: true });
+    expect(getCell(model, "A1")).toBe(undefined);
+  });
+
+  test("Cut and paste of a date format cell, by toggling formula visibility should return date", async () => {
```
Comment (rrahir) [inline tests/components/grid.test.ts:1445]: it is a broader behaviour than this, I just provided one example. if you look at the code it returns the formattedvalue as long as 1) the cell is not a formula, 2) we don't set 'formulaVisibility' to true. The test should be about 
1) 'cut a non formula cell returns its formatted value'
2) and another test 'cut a formula cell returns the content if `showFormulas === true` where you cut paste with showformulas with true and false values. 
The second test (or equivalent) might already exist though, I have not read the whole file.

Comment (khpa-odoo) [reply tests/components/grid.test.ts]: Pushed the changes :+1: 
Do checkout the previous comment as well.

**Review by rrahir — COMMENTED** [review]
last small comment, would be nice if you could add a test as well :)

**Comment by khpa-odoo** [comment]
> last small comment, would be nice if you could add a test as well :)

We do have a test case for 'onlyFormat' in case of `PASTE_FROM_OS_CLIPBOARD` in `menu_item_registry.test.ts`. Do you want me to edit it? Or write a new one for something else? Let me know :)

**Comment by rrahir** [comment]
> > last small comment, would be nice if you could add a test as well :)
> 
> We do have a test case for 'onlyFormat' in case of `PASTE_FROM_OS_CLIPBOARD` in `menu_item_registry.test.ts`. Do you want me to edit it? Or write a new one for something else? Let me know :)

I think another test would be nice:)

**Comment by khpa-odoo** [comment]
> I think another test would be nice:)
Added a new one :+1: 
Check it out. Thanks :)

**Review by rrahir — APPROVED** [review]
cheers :+1: 
@robodoo r+

**Comment by rrahir** [comment]
@robodoo rebase-ff
